### PR TITLE
New test harness for attributes test

### PR DESCRIPTION
**Pull request by Serabe**

### CODE REVIEW

**Comment by Serabe** [comment]
@chadhietala I would appreciate a first input on this. It is mostly the harness and the `renderTemplate` implementation was done mostly by imitation, so any input would be more than welcomed.

**Comment by Serabe** [comment]
#561

File: packages/@glimmer/runtime/test/attributes-test.ts [hunk 1]
```diff
@@ -68,204 +62,214 @@ function nativeValueForElementProperty<T extends keyof HTMLElementTagNameMap, P
...
 
-  equalTokens(root, '<input />');
-
-  rerender({ enabled: true });
-
-  equalTokens(root, '<input disabled />');
+  constructor(env = new TestEnvironment()) {
+    super(env);
+    this.element = env.getDOM().createElement('div') as HTMLDivElement;
+    //this.element.setAttribute('debug-root', 'true');
```
Comment (chadhietala) [inline packages/@glimmer/runtime/test/attributes-test.ts]: Remove ✂️

File: packages/@glimmer/runtime/test/attributes-test.ts [hunk 2]
```diff
@@ -68,204 +62,214 @@ function nativeValueForElementProperty<T extends keyof HTMLElementTagNameMap, P
...
-  rerender({ enabled: false });
-
-  equalTokens(root, '<input disabled />');
-
-  rerender({ enabled: 'wat' });
-
-  equalTokens(root, '<input disabled />');
-
-  rerender({ enabled: null });
+abstract class RenderTest extends BaseRenderTest {
```
Comment (chadhietala) [inline packages/@glimmer/runtime/test/attributes-test.ts]: I merged #560 so I'm not sure if you need this any more.

File: packages/@glimmer/runtime/test/attributes-test.ts [hunk 3]
```diff
@@ -68,204 +62,214 @@ function nativeValueForElementProperty<T extends keyof HTMLElementTagNameMap, P
...
-
-  rerender({ foo: 'http://foo.bar' });
+    this.rerender({ foo: 'javascript:foo()' });
+    this.assertHTML('<a href="unsafe:javascript:foo()"></a>');
+    this.assertStableNodes();
+  }
 
-  equalTokens(root, '<a href="http://foo.bar"></a>');
-});
+  @test "a[href] marks javascript: protocol as unsafe on updates"() {
```
Comment (chadhietala) [inline packages/@glimmer/runtime/test/attributes-test.ts]: This seems to be covered at [L190](https://github.com/glimmerjs/glimmer-vm/pull/564/files?diff=split#diff-37847c52bf54b71f8378543edd6d9407R190)

**Comment by Serabe** [comment]
@chadhietala this is ready for review. The tests related to sauce labs are failing (the four of them). Shall I just restart job like in Ember or is there anything I am doing wrong here?

OTOH, is there any way to remove the TS2300 and related errors?

Thank you!

**Comment by chadhietala** [comment]
@Serabe I think there might be legit errors here with IE. I think you might need to write some type of condition for IE like I did [here](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/@glimmer/runtime/test/initial-render-test.ts#L7-L21) looks like `textarea` doesn't behave the same. What is the `TS2300` error?

**Comment by chadhietala** [comment]
@Serabe this looks great otherwise.

**Comment by Serabe** [comment]
TS2300 and related errors are warnings from TypeScript when it finds a duplicated definition. It seems to be coming from the type declaration of es2015 and such. It raises so many that Travis log is hard to use.

I've added the select quirk and I'm looking for the one for textarea.

Thank you!

**Comment by Serabe** [comment]
0003screenshot

Seems that nodes are not stable in `"handles empty string textarea values"` after setting the value to `Alex`

**Comment by Serabe** [comment]
2017-07-11 at 16 59

The errors I mentioned were these. There are over 8.5K lines of this bloating travis log :$

**Comment by Serabe** [comment]
I've been a few hours digging into this in IE11 and cannot find anything :(

Any help available figuring this out?

**Comment by chadhietala** [comment]
@krisselden I check on a VM of IE and it appeared IE and Edge set do different things with CDATA in a textarea c/d?

**Comment by Serabe** [comment]
@chadhietala anything I can do to help fix this?

**Comment by Serabe** [comment]
@chadhietala @krisselden ping?

**Comment by chadhietala** [comment]
@serabe we removed IE9/IE10/Pantom from master. Can you try to rebase and see if the tests just pass?

**Comment by Serabe** [comment]
Done. Still failing IE11 and Edge.

**Comment by locks** [comment]
@Serabe You've Got Conflicts™.

**Comment by Serabe** [comment]
@locks already in Berlin, not sure when I'll be able to rebase again, but the underlying issue is still there.

**Comment by chadhietala** [comment]
Ok I have an answer. Firstly `value` is [not a valid html attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea) for `textarea`. Yet this still works if you set the prop, which is what we do today. That being said it appears that browsers differ how they represent the text. In Chrome you don't see a text node with the value instead you just see the following:

screen shot 2017-10-09 at 5 05 04 pm

Notice it does not have a `firstChild`. Now with IE and Edge if we start off with an empty string or `null` we have a similar behavior.

screen shot 2017-10-09 at 5 18 52 pm


However when we do update the `textarea` we have the following:

screen shot 2017-10-09 at 5 05 42 pm

So IE and Edge inject a TextNode when `value` is set.

**Bold What Shall We Do???**
Well because the result is the same between these browsers but the implementation details are slightly different we should just not `assertStableNodes` as they are not going to be stable. Instead we should just keep the assertion that the value updated. I can rebased and fix this if you don't have the time.

**Comment by chadhietala** [comment]
Merged as part of #698. Sorry this took so long @Serabe.

**Pull request closed without merging**